File: engine/src/main/java/de/hanno/hpengine/engine/model/IndexBuffer.java
package de.hanno.hpengine.engine.model;

import de.hanno.hpengine.engine.graphics.GpuContext;
import de.hanno.hpengine.engine.graphics.buffer.AbstractPersistentMappedBuffer;

import java.nio.IntBuffer;

import static org.lwjgl.opengl.GL15.GL_ELEMENT_ARRAY_BUFFER;

public class IndexBuffer extends AbstractPersistentMappedBuffer {

    public IndexBuffer(GpuContext gpuContext) {
        this(gpuContext, GL_ELEMENT_ARRAY_BUFFER);
        ensureCapacityInBytes(4*3*50);
    }

    public IndexBuffer(GpuContext gpuContext, int target) {
        super(gpuContext, target);
    }

    public IndexBuffer(GpuContext gpuContext, IntBuffer intBuffer) {
        this(gpuContext);
        put(intBuffer);
    }

    public void put(int[] values) {
        put(0, values);
    }
    public void put(int offset, int[] values) {
        ensureCapacityInBytes((values.length+offset)*Integer.BYTES);
        IntBuffer intBuffer = getBuffer().asIntBuffer();
        intBuffer.position(offset);
        intBuffer.put(values);
        getBuffer().rewind();
    }

    /**
     *
     * @param offset
     * @param nonOffsetIndices indices as if no other indices were before in the index buffer
     */
    public void appendIndices(int offset, int... nonOffsetIndices) {
        getBuffer().rewind();
        ensureCapacityInBytes((nonOffsetIndices.length+offset)*Integer.BYTES);
        if(offset == 0) {
            put(nonOffsetIndices);
        } else {
            for(int i = 0; i < nonOffsetIndices.length; i++) {
                put(offset+i, nonOffsetIndices[i]);
            }
        }
    }

    public void put(IntBuffer indices) {
        getBuffer().rewind();
        indices.rewind();
        getBuffer().asIntBuffer().put(indices);
    }

    public void put(int index, int value) {
        getBuffer().rewind();
        getBuffer().asIntBuffer().put(index, value);
    }

    public int getSize() {
        return getSizeInBytes() / Integer.BYTES;
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/vertexbuffer/IndexBuffer.kt
package de.hanno.hpengine.engine.vertexbuffer

import de.hanno.struct.copyTo
import de.hanno.hpengine.engine.graphics.GpuContext
import de.hanno.hpengine.engine.graphics.buffer.AbstractPersistentMappedBuffer
import java.nio.IntBuffer
import de.hanno.hpengine.engine.graphics.renderer.pipelines.IntStruct
import de.hanno.struct.StructArray
import org.lwjgl.opengl.GL15

open class IndexBuffer(gpuContext: GpuContext<*>?, target: Int) : AbstractPersistentMappedBuffer(
    gpuContext!!, target
) {
    constructor(gpuContext: GpuContext<*>?) : this(gpuContext, GL15.GL_ELEMENT_ARRAY_BUFFER) {
        ensureCapacityInBytes(4 * 3 * 50)
    }

    constructor(gpuContext: GpuContext<*>?, intBuffer: IntBuffer) : this(gpuContext) {
        put(intBuffer)
    }

    fun put(values: IntArray) {
        put(0, values)
    }

    fun put(offset: Int, values: IntArray) {
        ensureCapacityInBytes((values.size + offset) * Integer.BYTES)
        val intBuffer = buffer.asIntBuffer()
        intBuffer.position(offset)
        intBuffer.put(values)
        buffer.rewind()
    }

    /**
     *
     * @param offset
     * @param nonOffsetIndices indices as if no other indices were before in the index buffer
     */
    fun appendIndices(offset: Int, vararg nonOffsetIndices: Int) {
        buffer.rewind()
        ensureCapacityInBytes((nonOffsetIndices.size + offset) * Integer.BYTES)
        if (offset == 0) {
            put(nonOffsetIndices)
        } else {
            for (i in 0 until nonOffsetIndices.size) {
                put(offset + i, nonOffsetIndices[i])
            }
        }
    }

    fun put(indices: IntBuffer) {
        buffer.rewind()
        indices.rewind()
        buffer.asIntBuffer().put(indices)
    }

    fun put(index: Int, value: Int) {
        buffer.rewind()
        buffer.asIntBuffer().put(index, value)
    }

    val size: Int
        get() = sizeInBytes / Integer.BYTES

    fun appendIndices(indexOffset: Int, indices: StructArray<IntStruct>) {
        buffer.rewind()
        ensureCapacityInBytes((indexOffset + indices.size) * Integer.BYTES)
        indices.buffer.copyTo(buffer, true, indexOffset * Integer.BYTES)
    }
}